File: Sprint4/A/C++/main.cpp
// https://contest.yandex.ru/contest/23991/problems/A/
#include <iostream>

long long poly_hash(int a, int m, std::string& s) {
    long long h = 0, power = 1;
    int n = int(s.length());
    for (int i = n - 1; i >= 0; --i) {
        h = (h + int(s[i]) * power) % m;
        power = (power * a) % m;
    }
    return h;
}

int main() {
    int a, m;
    std::string s;

    std::cin >> a >> m;
    std::cin >> s;

    std::cout << poly_hash(a, m, s) << std::endl;
}

File: Sprint4/A/Rust/src/main.rs
// https://contest.yandex.ru/contest/23991/problems/A/
use std::io::BufRead;

fn poly_hash(a: usize, m: usize, s: String) -> u64 {
    let mut h = 0u64;
    let mut power = 1u64;

    for ch in s.chars().rev() {
        h = (h + ch as u64 * power) % m as u64;
        power = (power * a as u64) % m as u64;
    }

    h
}
fn main() {
    let stdin = std::io::stdin();
    let mut lines = stdin.lock().lines();

    let a: usize = lines.next().unwrap().unwrap().parse().unwrap();
    let m: usize = lines.next().unwrap().unwrap().parse().unwrap();

    let s = lines.next().unwrap().unwrap();

    println!("{}", poly_hash(a, m, s));
}
